File: ast.go
// Copyright 2015 Google Inc. All rights reserved
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package kati

import (
	"bytes"
	"strings"
)

type ast interface {
	eval(*Evaluator) error
	show()
}

type assignAST struct {
	srcpos
	lhs Value
	rhs Value
	op  string
	opt string // "override", "export"
}

func (ast *assignAST) eval(ev *Evaluator) error {
	return ev.evalAssign(ast)
}

func (ast *assignAST) evalRHS(ev *Evaluator, lhs string) (Var, error) {
	origin := "file"
	if ast.filename == bootstrapMakefileName {
		origin = "default"
	}
	if ast.opt == "override" {
		origin = "override"
	}
	// TODO(ukai): handle ast.opt == "export"
	switch ast.op {
	case ":=":
		switch v := ast.rhs.(type) {
		case literal:
			return &simpleVar{value: v.String(), origin: origin}, nil
		case tmpval:
			return &simpleVar{value: v.String(), origin: origin}, nil
		default:
			var buf bytes.Buffer
			err := v.Eval(&buf, ev)
			if err != nil {
				return nil, err
			}
			return &simpleVar{value: buf.String(), origin: origin}, nil
		}
	case "=":
		return &recursiveVar{expr: ast.rhs, origin: origin}, nil
	case "+=":
		prev := ev.lookupVarInCurrentScope(lhs)
		if !prev.IsDefined() {
			return &recursiveVar{expr: ast.rhs, origin: origin}, nil
		}
		return prev.AppendVar(ev, ast.rhs)
	case "?=":
		prev := ev.lookupVarInCurrentScope(lhs)
		if prev.IsDefined() {
			return prev, nil
		}
		return &recursiveVar{expr: ast.rhs, origin: origin}, nil
	}
	return nil, ast.errorf("unknown assign op: %q", ast.op)
}

func (ast *assignAST) show() {
	logf("%s %s %s %q", ast.opt, ast.lhs, ast.op, ast.rhs)
}

// maybeRuleAST is an ast for rule line.
// Note we cannot be sure what this is, until all variables in |expr|
// are expanded.
type maybeRuleAST struct {
	srcpos
	expr      Value
	term      byte // Either ':', '=', or 0
	afterTerm []byte
}

func (ast *maybeRuleAST) eval(ev *Evaluator) error {
	return ev.evalMaybeRule(ast)
}

func (ast *maybeRuleAST) show() {
	logf("%s", ast.expr)
}

type commandAST struct {
	srcpos
	cmd string
}

func (ast *commandAST) eval(ev *Evaluator) error {
	return ev.evalCommand(ast)
}

func (ast *commandAST) show() {
	logf("\t%s", strings.Replace(ast.cmd, "\n", `\n`, -1))
}

type includeAST struct {
	srcpos
	expr string
	op   string
}

func (ast *includeAST) eval(ev *Evaluator) error {
	return ev.evalInclude(ast)
}

func (ast *includeAST) show() {
	logf("include %s", ast.expr)
}

type ifAST struct {
	srcpos
	op         string
	lhs        Value
	rhs        Value // Empty if |op| is ifdef or ifndef.
	trueStmts  []ast
	falseStmts []ast
}

func (ast *ifAST) eval(ev *Evaluator) error {
	return ev.evalIf(ast)
}

func (ast *ifAST) show() {
	// TODO
	logf("if")
}

type exportAST struct {
	srcpos
	expr   []byte
	export bool
}

func (ast *exportAST) eval(ev *Evaluator) error {
	return ev.evalExport(ast)
}

func (ast *exportAST) show() {
	// TODO
	logf("export")
}

File: ast.cc
// Copyright 2015 Google Inc. All rights reserved
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// +build ignore

#include "ast.h"

#include "eval.h"
#include "stringprintf.h"
#include "strutil.h"
#include "value.h"

AST::AST() {}

AST::~AST() {}

string RuleAST::DebugString() const {
  return StringPrintf("RuleAST(expr=%s term=%d after_term=%s loc=%s:%d)",
                      expr->DebugString().c_str(),
                      term,
                      after_term->DebugString().c_str(),
                      LOCF(loc()));
}

string AssignAST::DebugString() const {
  const char* opstr = "???";
  switch (op) {
    case AssignOp::EQ: opstr = "EQ"; break;
    case AssignOp::COLON_EQ: opstr = "COLON_EQ"; break;
    case AssignOp::PLUS_EQ: opstr = "PLUS_EQ"; break;
    case AssignOp::QUESTION_EQ: opstr = "QUESTION_EQ"; break;
  }
  const char* dirstr = "???";
  switch (directive) {
    case AssignDirective::NONE: dirstr = ""; break;
    case AssignDirective::OVERRIDE: dirstr = "override"; break;
    case AssignDirective::EXPORT: dirstr = "export"; break;
  }
  return StringPrintf("AssignAST(lhs=%s rhs=%s (%s) "
                      "opstr=%s dir=%s loc=%s:%d)",
                      lhs->DebugString().c_str(),
                      rhs->DebugString().c_str(),
                      NoLineBreak(orig_rhs.as_string()).c_str(),
                      opstr, dirstr, LOCF(loc()));
}

string CommandAST::DebugString() const {
  return StringPrintf("CommandAST(%s, loc=%s:%d)",
                      expr->DebugString().c_str(), LOCF(loc()));
}

string IfAST::DebugString() const {
  const char* opstr = "???";
  switch (op) {
    case CondOp::IFEQ: opstr = "ifeq"; break;
    case CondOp::IFNEQ: opstr = "ifneq"; break;
    case CondOp::IFDEF: opstr = "ifdef"; break;
    case CondOp::IFNDEF: opstr = "ifndef"; break;
  }
  return StringPrintf("IfAST(op=%s, lhs=%s, rhs=%s t=%zu f=%zu loc=%s:%d)",
                      opstr,
                      lhs->DebugString().c_str(),
                      rhs->DebugString().c_str(),
                      true_asts.size(),
                      false_asts.size(),
                      LOCF(loc()));
}

string IncludeAST::DebugString() const {
  return StringPrintf("IncludeAST(%s, loc=%s:%d)",
                      expr->DebugString().c_str(), LOCF(loc()));
}

string ExportAST::DebugString() const {
  return StringPrintf("ExportAST(%s, %d, loc=%s:%d)",
                      expr->DebugString().c_str(),
                      is_export,
                      LOCF(loc()));
}

string ParseErrorAST::DebugString() const {
  return StringPrintf("ParseErrorAST(%s, loc=%s:%d)",
                      msg.c_str(),
                      LOCF(loc()));
}

RuleAST::~RuleAST() {
  delete expr;
  delete after_term;
}

void RuleAST::Eval(Evaluator* ev) const {
  ev->EvalRule(this);
}

AssignAST::~AssignAST() {
  delete lhs;
  delete rhs;
}

void AssignAST::Eval(Evaluator* ev) const {
  ev->EvalAssign(this);
}

CommandAST::~CommandAST() {
  delete expr;
}

void CommandAST::Eval(Evaluator* ev) const {
  ev->EvalCommand(this);
}

IfAST::~IfAST() {
  delete lhs;
  delete rhs;
}

void IfAST::Eval(Evaluator* ev) const {
  ev->EvalIf(this);
}

IncludeAST::~IncludeAST() {
  delete expr;
}

void IncludeAST::Eval(Evaluator* ev) const {
  ev->EvalInclude(this);
}

ExportAST::~ExportAST() {
  delete expr;
}

void ExportAST::Eval(Evaluator* ev) const {
  ev->EvalExport(this);
}

ParseErrorAST::~ParseErrorAST() {
}

void ParseErrorAST::Eval(Evaluator* ev) const {
  ev->set_loc(loc());
  ev->Error(msg);
}
